### DISCUSSION TITLE
Caching the Analysis phase to accelerate repeated builds (are there other ways? did I do something evil?)

### DISCUSSION THREAD

**Discussion by SnoopJ** [pyinstaller/pyinstaller#7626] [PyInstaller]
I've been doing some extension work on a PyInstaller application I maintain that requires a lot of re-building as I go, and the build process is quite long: about 30 minutes, 20 of which is spent during the `Analysis` phase.[1]

Today as I launched an additional build to deal with a silly typo and wished I could have a faster build, I realized I could cut down on that time appreciably by "just" caching the results of the `Analysis` phase and putting in a feature flag in my `.spec` file to use cached TOCs instead of re-walking the large graph.

I implemented a [proof of concept](https://github.com/SnoopJ/playground/tree/main/python/pyinstaller_/cached_analysis_phase) of this with `numpy` as a lone dependency, but as I put the final touches on that, I started wondering if I'd done something uncouth, or maybe there was a PyInstaller feature I'm not aware of that I wasn't taking proper advantage of.[2]

So I decided to post here. I welcome any and all feedback on the idea/proof of concept, or suggestions of other ways to cut down on repeated work when I'm building PyInstaller distributions frequently.

---

[1] This is in part because the application is large (main TOC is ~10k entries), and in part because I apply the `collect_all()` hook very aggressively. Rest assured, I've been [properly discouraged](https://github.com/pyinstaller/pyinstaller/issues/6785#issuecomment-1112662156) from doing it this way, but it keeps some maintenance headaches away.

[2]  My builds _do_ use `--clean`, but if I understand right, the result of the `Analysis` phase is not cached by PyInstaller because in general the graph may have changed, and cache invalidation is hard :sweat_smile:

**Comment by bwoodsend** [comment]
> the build process is quite long: about 30 minutes, 20 of which is spent during the Analysis phase.[

Running in a minimal virtual environment should help if you're not already doing that. That would at least limit the viral nature of `collect_all()`.

>  My builds do use --clean, but if I understand right, the result of the Analysis phase is not cached by PyInstaller because in general the graph may have changed, and cache invalidation is hard sweat_smile

The graph is cached but it's invalidated if any modules are changed since those changes may include changes imports. In practice, the cache will only get reused if you only add/change an `--add-data` or `--icon` flag. In a happier world, analysis caching could probably be done on a per-module basis (i.e. record that `numpy/__init__.py` with sha256sum `343ecee39666e2a9f81f24a223a6980e6aa7c853e3b615548f7b9317d7884a9e` imports `builtins • numpy • numpy.__config__ • numpy._distributor_init • numpy._globals • numpy._pytesttester • numpy.amax • numpy.amin • numpy.array • numpy.array_api • numpy.bool_ • numpy.bytes_ • numpy.compat • numpy.complexfloating • numpy.core • numpy.core._dtype_ctypes • numpy.ctypeslib • numpy.datetime64 • numpy.dtype • numpy.expand_dims • numpy.eye • numpy.fft • numpy.float32 • numpy.float64 • numpy.floating • numpy.generic • numpy.histogramdd • numpy.int16 • numpy.int32 • numpy.int64 • numpy.int8 • numpy.integer • numpy.iscomplexobj • numpy.lib • numpy.linalg • numpy.ma • numpy.matrixlib • numpy.ndarray • numpy.number • numpy.object_ • numpy.polynomial • numpy.random • numpy.recarray • numpy.str_ • numpy.testing • numpy.timedelta64 • numpy.ufunc • numpy.uint16 • numpy.uint32 • numpy.uint64 • numpy.uint8 • numpy.unsignedinteger • numpy.version • numpy.void • os • pathlib • sys • warnings`) so that minimal changes to code cause minimal rebuilding (note that traversing the graphs once you've scanned all the modules and ran all the corresponding hooks is quick). But the dependency scanning code is overcomplicated enough as it is and, like you say, cache invalidation is hard. It's particularly nasty to test and if there's anything wrong with it, it won't be caught by CI/CD and would go onto to create stateful issues that can't easily be reproduced and given [how much of a grind it can be](https://github.com/pyinstaller/pyinstaller/issues/7618) to get even trivially reproducible issues out of people, I expect I'd end up reverting such a feature out of anger.

**Reply by SnoopJ** [reply]
Yea unfortunately the venv is already as minimal as it can be. This is an application that has several of the enormous neural net frameworks in it, so it's just a big graph.

The fact that you didn't recoil in horror from my idea to do my own caching at least makes me feel better about it! I haven't folded this into the real application yet, will see if the wheels come off when I do that. Thanks also for the clarification about when the cache is going to apply in practice. I figured this would probably be something that has already been considered and decided against in the library itself, although I'd be happy to send a PR explaining the ad-hoc approach if there's any interest.

**Reply by bwoodsend** [reply]
> although I'd be happy to send a PR explaining the ad-hoc approach if there's any interest.

Stick around on our issue tracker long enough and you'll start to become vary wary of leaving users any rope that they could feasibly hang themselves with. I anticipate users looking for ways to speed up their builds, blindly copy/pasting the code and not reading about what it does then reporting that PyInstaller is broken because it won't pick up the dependency they just added.

**Comment by SnoopJ** [comment]
> I haven't folded this into the real application yet, will see if the wheels come off when I do that.

For the sake of leaving a complete record, the main thing I didn't account for in my initial proof-of-concept is `base_library.zip`, which needs to be cached separately for a build that uses `--clean`. I decided to 'just' copy the archive into my ad-hoc cache, although a more graceful caching solution that serializes the actual `PyiModuleGraph` could probably recreate the archive as needed. I went with the ad-hoc copy because the graph-oriented solution seems like it would require re-implementing large portions of `Analysis`.

So far, that's the only thing that's blown up on me in the 'real' application, we'll see how it goes. :crossed_fingers:
